django/contrib/postgres/fields/ranges.py

```diff
@@ -332,6 +341,15 @@ class RangeEndsWith(models.Transform):
     def output_field(self):
         return self.lhs.output_field.base_field
 
+    def as_postgresql(self, compiler, connection):
+        sql, params = self.as_sql(compiler, connection)
+        # Cast the lhs if needed.
+        cast_sql = ""
+        if isinstance(self.lhs, models.Expression) and self.lhs._output_field_or_none:
+            cast_internal_type = self.lhs.output_field.base_field.get_internal_type()
+            cast_sql = "::{}".format(connection.data_types.get(cast_internal_type))
```

Comment:
What do you think about using `get_placeholder()` instead?

Reply:
`get_placeholder()` is much better!  I didn't realise this was a thing as it's not in the docs (a separate PR to document it might be worth a go too). After doing some more testing I found that a lot of constraint validations on ranges were failing - not just with NULLs – because range literals are declared with strings and using `get_placeholder()` covers all cases.

I saw the `ArrayField.get_placeholder()` doing exactly what's needed so copied that over to `RangeField` and my tests all passed.

I've added your test + a test verifying that it works with NULLs as originally reported + a test to confirm `ArrayField` also works as array literals are also declared with strings.  I'm thinking the NULL test may no longer be needed though - would you prefer it to be removed?

I've reverted the original commit and appended a new one. Let me know when you'd like the original + revert dropped.